### DISCUSSION TITLE
Issue with model binding in v3

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#6952] [1. Help]
Before we properly switch everything over to v3, we need the old v2 functionality regarding models. I've set `legacy_model_binding` in our config, but we still have issues with casts on those models. We get things like:

```
Property type not supported in Livewire for property: ["pending"]
```

What can we do about this? I know the best thing is to move data into arrays but the upgrade PR is already massive, so it's preferred to do this in new PR's over time (it would be very time consuming to do this all now).

Edit: this appears to be coming from validation rules. We still use Spatie's Enum package, so we validation rules like the one below:

```php
'post.type' => new EnumRule(Type::class),
```

I will raise this with Spatie as well, because for some reason their code is not even called at this point. For those interested, a sample repository is available here: https://github.com/ju5t/livewire-issue-enums/commit/6c824556eac761661c65edd4e64919b59c6bfb51

**Comment by ju5t** [comment]
@joshhanley sorry for tagging you, but this appears to be a blocking change for our upgrade. Is this expected behaviour or is this a bug somewhere?

**Reply by lairg99** [reply]
That sounds weird. Could you provide a code example?

**Comment by lairg99** [comment]
I can't reproduce your error. Could you provide some details and an excerpt of your code? Moreover, I would like to mention my discussion about this topic, because in my eyes it's a feature and doesn't need to be replaced. See #6909

```php
class Update extends Component
{
    public Post $post;

    protected function rules(): array {
        return ['post.name' => ['string']];
    }

    public function render() {
        return view('post.modal.create');
    }
}
```

```html
<div>
    <div>{{ $post->name }}</div>

    <input type="text" wire:model.live="post.name"/>
</div>
```

**Reply by ju5t** [reply]
@lairg99 yea this appears to be coming from spatie/laravel-enum and/or Livewire. I've shared a test-repo in my original post.

**Reply by lairg99** [reply]
Well, I read over it that it has to do with enums. I tried it with native enums and everything works fine. Is there any special reason _not_ to use native enums? 

However, I can reproduce your error with _spatie/laravel-enum_. If the support of Spatie's enums is required, take a look at [synthesizers](https://livewire.laravel.com/docs/synthesizers). The given exception occurs because there's no synthesizers that matches the type. I assume that v2 tried to cast it as a string(?)

```php
protected function getSynthesizerByTarget($target, $context, $path)
{
    foreach ($this->propertySynthesizers as $synth) {
        if ($synth::match($target)) {
            return new $synth($context, $path);
        }
    }

    throw new \Exception('Property type not supported in Livewire for property: ['.json_encode($target).']');
}
``` 

You can extend Livewire's EnumSynth so that it also matched with Spatie's Enums:

```php
namespace App\Synthesizers;

use Livewire\Mechanisms\HandleComponents\Synthesizers\EnumSynth as BaseEnumSynth;
use Spatie\Enum\Enum;

class EnumSynth extends BaseEnumSynth
{
    public static $key = 'spatie-enum';

    static function match($target): bool {
        return $target instanceof Enum;
    }
}
```

In addition, you have register it in your service provider.

```php
Livewire::propertySynthesizer(EnumSynth::class);
```

Then everything should work as expected.

**Reply by ju5t** [reply]
It surprises me that this would need a different synthesizer. As we're using it as a property on `Post`, I expected to be done by the cast which we configured on the model. We don't use Spatie's enums as a public property on the component.

So apparently `legacy_model_binding` is not 100% backwards compatible. That's a bit unexpected.

**Reply by ju5t** [reply]
Adding a synthesizer solves it by the way, but we have a similar issue with Spatie's model states for example. There might be more. This will be an interesting upgrade.

**Reply by lairg99** [reply]
> I expected to be done by the cast which we configured on the model.

You're right, that's a valid point. As I mentioned in #6909, I'm pondering whether model binding is a legacy concept or a feature. In our applications we rely heavily on this feature - I would appreciate your input in this discussion.
